promoted_python_delivery_client/client/sdk_delivery.py

```diff
@@ -14,16 +13,17 @@ def __init__(self) -> None:
     def run_delivery(self, request: DeliveryRequest) -> Response:
         req = request.request
         paging = req.paging
-        if paging is not None and paging.offset is not None and paging.offset >= len(req.insertion):
-            raise ValueError("invalid paging (offset >= size)")
 
         # Set a request id.
         req.request_id = str(uuid.uuid4())
         if paging is None:
             paging = Paging(offset=0, size=len(req.insertion))
 
+        # Adjust size and offset.
         offset = max(0, paging.offset) if paging.offset is not None else 0
-        index = offset if request.insertion_page_type != InsertionPageType.PREPAGED else 0
+        if offset < request.insertion_start:
+            raise ValueError("offset should be >= insertion start (specifically, the global position)")
+        index = offset - request.insertion_start
```

Comment:
This looks good.

What happens if `offset + paging.size >= request.insertion_start + len(req.insertion)` ?   It looks good (truncates using the final_insertion_size logic).  I just want to double check.

Reply:
That's the same case as test_insertion_start_less_than_offset unless I'm missing something. (These two tests, I ported directly from the TS client.)